<?xml version="1.0"?>
<launch>
    <!-- TransBot GUI Control Launch File -->
    <!-- Pure GUI control with direct joint angle arm control -->
    
    <!-- Parameters -->
    <param name="robot_name" value="transbot" />
    <param name="use_sim_time" value="false" />
    
    <!-- Camera parameters -->
    <arg name="camera_device" default="0" />
    <arg name="camera_width" default="640" />
    <arg name="camera_height" default="480" />
    <arg name="camera_fps" default="10" />
    
    <!-- Arm parameters -->
    <arg name="use_arm" default="true" />
    <arg name="arm_simulation" default="false" />
    
    <!-- Base control parameters -->
    <arg name="max_linear_vel" default="1.0" />
    <arg name="max_angular_vel" default="2.0" />
    
    <!-- 1. Base driver -->
    <include file="$(find transbot_bringup)/launch/bringup.launch" />
    
    <!-- 2. Camera node for line following -->
    <node name="usb_cam" pkg="usb_cam" type="usb_cam_node" output="screen">
        <param name="video_device" value="/dev/video$(arg camera_device)" />
        <param name="image_width" value="$(arg camera_width)" />
        <param name="image_height" value="$(arg camera_height)" />
        <param name="pixel_format" value="yuyv" />
        <param name="camera_frame_id" value="usb_cam" />
        <param name="io_method" value="mmap"/>
        <param name="framerate" value="$(arg camera_fps)" />
        <!-- 直接发布到transbot_linefollow期望的话题 -->
    </node>
    
    <!-- 2.1 Camera Controller - controls camera servo angle -->
    <node name="camera_controller" pkg="transbot_composite_app" 
          type="camera_controller.py" output="log" respawn="true">
    </node>
    
    <!-- 3. Line following node - let it handle its own camera -->
    <node name="LineDetect" pkg="transbot_linefollow" type="follow_line.py" output="screen">
        <!-- VideoSwitch=false 表示显示OpenCV调试窗口 -->
        <param name="VideoSwitch" type="bool" value="false"/>
        <param name="img_flip" type="bool" value="false"/>
        <!-- 重新映射cmd_vel话题，让控制器来决定是否转发 -->
        <remap from="/cmd_vel" to="/cmd_vel_linefollow"/>
    </node>
    
    <!-- 2.3 Simple Line Follow Controller - controls when linefollow commands are forwarded -->
    <node name="simple_linefollow_controller" pkg="transbot_composite_app" 
          type="simple_linefollow_controller.py" output="screen" respawn="true">
        <param name="package_path" value="/home/pi/src/transbot_ws/src/transbot_composite_app" />
    </node>
    
    <!-- 3. Main controller -->
    <node name="composite_main_controller" pkg="transbot_composite_app" 
          type="main.py" output="log" respawn="true">
        <param name="max_linear_vel" value="$(arg max_linear_vel)" />
        <param name="max_angular_vel" value="$(arg max_angular_vel)" />
    </node>
    
    <!-- 4. Line follow controller -->
    <!-- 注释掉旧的巡线控制器，现在使用简单控制器 -->
    <!--
    <node name="linefollow_controller" pkg="transbot_composite_app" 
          type="linefollow_controller.py" output="screen" respawn="true">
        <param name="camera_topic" value="/camera/image_raw" />
        <param name="debug_image" value="true" />
        <rosparam file="$(find transbot_composite_app)/config/linefollow_params.yaml" command="load" ns="linefollow_controller" />
    </node>
    -->
    
    <!-- 5. Direct Joint Angle Arm Controller -->
    <group if="$(arg use_arm)">
        <node name="arm_controller" pkg="transbot_composite_app" 
              type="arm_controller.py" output="screen" respawn="true">
            <param name="simulation_mode" value="$(arg arm_simulation)" />
            <param name="move_speed" value="0.5" />
            <param name="interpolation_steps" value="20" />
        </node>
    </group>
    
    <!-- 6. GUI Control Panel -->
    <node name="gui_controller" pkg="transbot_composite_app" 
          type="gui_server.py" output="screen" respawn="false" required="true">
        <param name="debug_mode" value="false" />
        <param name="video_fps" value="10" />
        <param name="window_title" value="TransBot Control Panel" />
        <param name="linefollow_debug" value="true" />
    </node>
    
    <!-- 7. Static transforms -->
    <node pkg="tf2_ros" type="static_transform_publisher" name="base_to_camera"
          args="0.1 0 0.3 0 0 0 base_link camera_link" />
    
    <!-- 8. Startup message -->
    <node name="startup_message" pkg="transbot_composite_app" type="startup_message.py" output="screen" />
    
</launch> 